### ISSUE TITLE
Incorrect Jacobian of `jax.scipy.special.logsumexp` when `b` contains a `0`

### ISSUE THREAD

**Issue by Adrien-Kahn** [jax-ml/jax#22398]
### Description

The gradient of `jax.scipy.special.logsumexp` along `b` is wrong when `b` contains a `0`. In the following example, the gradient of $f(x_1, x_2) = \log(x_1 + x_2)$ should be:
```math
\nabla_x f(x) = \left( \frac{1}{x_1 + x_2}, \frac{1}{x_1 + x_2} \right)
```
but JAX outputs that the gradient at `(1,0)` is `(1,0)` when it should be `(1,1)`:

```python
import jax
import jax.numpy as jnp

fun = lambda x: jax.scipy.special.logsumexp(jnp.zeros(2), axis=0, b=x)
a = jnp.array([1, 0], dtype=float)

jax.jacfwd(fun)(a)
```
output:
```python
Array([1., 0.], dtype=float32)
```

There is only a problem for an exact `0`: the gradient taken at `(1,1e-30)` outputs the correct value of `(1,1)`:
```python
a = jnp.array([1, 1e-30], dtype=float)
jax.jacfwd(fun)(a)
```
output:
```python
Array([1., 1.], dtype=float32)
```

Furthermore, if the `logsumexp` is implemented manually, the gradient at `(1, 0)` is also correct:
```python
fun_2 = lambda x: jnp.log(jnp.sum(x * jnp.exp(jnp.zeros(2))))
a = jnp.array([1, 0], dtype=float)
jax.jacfwd(fun_2)(a)
```
output:
```python
Array([1., 1.], dtype=float32)
```

cc @PhilipVinc

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.30
jaxlib: 0.4.30
numpy:  1.24.4
python: 3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 10:37:07) [Clang 15.0.7 ]
jax.devices (1 total, 1 local): [CpuDevice(id=0)]
process_count: 1
platform: uname_result(system='Darwin', node='mba-10836208', release='23.4.0', version='Darwin Kernel Version 23.4.0: Fri Mar 15 00:19:22 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T8112', machine='arm64')
```

**Comment by jakevdp** [comment]
Hi - thanks for the clear report! Indeed this looks like a bug in the `logsumexp` implementation; I think the root cause is this line: https://github.com/google/jax/blob/46103f6ff3f4cbac2881fece37d145bba7d3e60d/jax/_src/ops/special.py#L74

I suspect the best way of fixing this will be to add a custom JVP rule.

**Comment by NDOWAH** [comment]
Hello @superbobry, though a beginner, i would like to give this a look to make my first contribution. Any documentation or idea to get me started would be appreciated. I have a clear understanding of how jacobian is being computed.

**Comment by jakevdp** [comment]
Reopening because the fix had to be rolled back

**Comment by PhilipVinc** [comment]
What was the problem?

**Comment by jakevdp** [comment]
The change led to failing tests in other projects.

**Comment by marcocuturi** [comment]
not a contribution.
could this implementation
https://github.com/ott-jax/ott/blob/561adbbda9a88ea11d42d3262dbb5ce81bd482e8/src/ott/math/utils.py#L142
be useful?

**Comment by adityasingh2400** [comment]
Cross-linking for triage: this appears to be the same root cause as #37633, and there is an open PR addressing it at #37729 (currently in review, with feedback from the Gemini reviewer about narrowing the NaN mask in `jax/_src/ops/special.py`).

Re-derived the math to sanity check the expected output in the original report. For `logsumexp(a, b) = log(sum_i b_i exp(a_i))`, the partial with respect to `b_i` is `exp(a_i) / sum_j b_j exp(a_j)`, which does not vanish at `b_i = 0`. With `a = (0, 0)` and `b = (1, 0)` the derivative is `(1, 1)`, matching the report. The current behavior returns `(1, 0)` because the implementation substitutes `a_i <- -inf` at `b_i = 0` positions (to keep `amax` finite when `a` can be `+inf` there), and that substitution kills gradient flow through `exp(a - amax)` with respect to `b`.

The fix in #37729 splits the substitution so it only feeds `amax` and leaves the unsubstituted `a` in the `exp(a - amax)` path. This preserves existing behavior whenever `b > 0` everywhere and only changes the `b == 0` contribution, which should keep it a strict superset of the path that was rolled back from #22836 / #22879. Worth landing once the NaN mask scoping is tightened per the review.
